### DISCUSSION TITLE
How to convert YUV420 video format input and use in media element (WPF - C#)

### DISCUSSION THREAD

**Discussion by kondareddy** [ffmpeginteropx/FFmpegInteropX#334] [Q&A]
We are integrating Agora in WPF (C#) - Agora SDK gives use raw video frames - which we need to render on WPF screens.

SDK Documentation Link.

https://docs-preprod.agora.io/en/video-legacy/API%20Reference/csharp/API/rtc_api_data_type.html#enum_videoframetype

We are searching for library which takes YUV420 format and give use stream, which can be used as media element source.

https://learn.microsoft.com/en-us/dotnet/desktop/wpf/graphics-multimedia/how-to-control-a-mediaelement-play-pause-stop-volume-and-speed?view=netframeworkdesktop-4.8

**Comment by brabebhin** [comment]
Hi,

Unfortunetely WPF media element is quite limited in this regard. Your only hope would be to implement a media source with old MF interfaces.

https://learn.microsoft.com/en-us/windows/win32/api/mfidl/nn-mfidl-imfmediasource

However, if using MediaElement is not a requirement, you can use the MediaPlayerElement wrapped control for desktop applications

https://learn.microsoft.com/en-us/windows/communitytoolkit/controls/wpf-winforms/mediaplayerelement

The MediaPlayerElement is much more extensible. 
You can also skip the agora SDK and use this SDK- we can handle most audio and video streams out of the box with a rather simple API. My recommendation is to do so: using C# for media playerback typically doesn't ensure a smooth experience, due to GC delays. Our SDK is full native C++ (you can use it safely from desktop C# with a nuget package)

If the above is not good enough, you can use the winRT MediaPlayer in frame server mode, with a custom built MediaStreamSource of your own. You will need to handle rendering to a DirectX surface though

https://learn.microsoft.com/en-us/windows/uwp/audio-video-camera/play-audio-and-video-with-mediaplayer#use-mediaplayer-in-frame-server-mode

https://msdn.microsoft.com/library/windows/apps/windows.media.core.mediastreamsource

Windows might be able to render the YUV directly if you provide the proper encoding properties in the sample metadata.

However, I stromgly recommend you use ffmpeginteropx instead :)

**Comment by lukasf** [comment]
If you can use MediaPlayerElement (either by creating a WinUI app, or by hosting WinUI MediaPlayerElement in a [WPF XamlIsland](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/xaml-islands)), then you should be able to easily render your YUV420 video using MediaStreamSource, without any additional frameworks (also without FFmpegInteropX).

You just need to create VideoEncodingProperties with MediaEncodingSubtypes.Iyuv, and create a MSS from that, somehow like this (minimal setup):

```
var descriptor = new VideoStreamDescriptor(VideoEncodingProperties.CreateUncompressed(MediaEncodingSubtypes.Iyuv, witdth, height));
var mss = new MediaStreamSource(descriptor);
mss.SampleRequested += ...
var mediaPlaybackItem = new MediaPlaybackItem(mss);
```

In the SampleRequested event handler, create a MediaStreamSample, fill the Sample's Buffer with the raw YUV420 data and set Timestamp, and set the sample to the eventArgs. Now you should be able to assign the mediaPlaybackItem to MediaPlayer.Source (if you want to use MediaPlayerElement with MediaPlayer, which is recommended), or directly to MediaPlayerElement.

Note that it is recommended to handle SampleRequested in native code, to avoid delay due to GC. So recommendation would be to do this in a C++/WinRT or C++/CX component, and use that from C#. But you can first try easily in C# to see if it generally works with your data. If you want to use it and want 100% stable frame rates, consider moving to native code.

Feel free to check our code for comparison, especially UncompressedVideoSampleProvider and FFmpegMediaSource.

**Reply by MayankKr10** [reply]
Our fps range is from 5-30, how can we continuously create source from the received byte array and play it as a video?
I tried this out but sometimes the frames are visible and sometimes they are not (changing mostly according to the TimeStamp). Is there something wrong/missing in my implementation:

**Sample callback handling**
var descriptor = new VideoStreamDescriptor(VideoEncodingProperties.CreateUncompressed(MediaEncodingSubtypes.Iyuv, 848, 480));
            var mss = new MediaStreamSource(descriptor);
            var arr1 = File.ReadAllBytes(@"C:\\Users\\Mayank4.Kumar\\Pictures\\video.yuv"); // sample file for rendering YUV data
            mss.SampleRequested += (sender, e) => MediaSampleRequested(arr1.AsBuffer(), e);
            VideoPlayer.MediaPlayer.Source = MediaSource.CreateFromMediaStreamSource(mss);


 private void MediaSampleRequested(IBuffer buffer, MediaStreamSourceSampleRequestedEventArgs args)
        {
            var streamSample = MediaStreamSample.CreateFromBuffer(buffer, new TimeSpan(1000000));
            args.Request.Sample = streamSample;
            args.Request.Sample.DecodeTimestamp = new TimeSpan(1000000);
        }

What is the suggested TimeSpan for a real-time video for around 15fps as mentioned above?

The memory usage also seems to shoot up in this case (tried in C# for now) and it keeps on increasing:
![image](https://user-images.githubusercontent.com/48569729/212290713-f0419345-0195-497f-845f-b6165ef791a4.png)

**Reply by brabebhin** [reply]
Media Foundation does freak out if you send the wrong timestamp and wrong sample length, it essentially buffers samples but doesn't consume them, thus sending the memory usage to the skies.

The timestap depends on your resolution, bit rate and fps. So you should take into account the amount of data that you have in the buffer. You essentially need to compute how much data you have in your file. The first time stamp will always be 0. The second timestamp will be 0+ length of first sample. Third one will be timestamp + length of first sample + length of second sample and so on. You need to handle these calculations properly, they are absolutely vital for good playback. We had lots of bugs with edge cases in which we didn't compte those correctly and all sorts of weird things were happening.

timestamp = 0;
First sample: timestam =0, length = length of first sample
timestamp = timestamp + length(First sample)
...

In other words, update the timestamp before each sample with the length of the previous sample.

^^ the above asumes there is no seeking in the stream. If there's seeking, you need to adjust the timestamp based on seek position.

The size of a single frame should be width * height * bitrate (or bits per pixel) (assuming you have full frames). 
Based on this you need to see how many frames you have in your files.
If your fps is 15 frames, you can then compute timestamps and length of samples.

If your framerate is variable, I am affraid things get a little more complicated. MediaStreamSource assumes the FPS is the same. If you cannot provide enough frames, from your stream, you either need to provide mock frames to reach the desired FPS, or playback will not be real time.

**Reply by lukasf** [reply]
For getting your test app working: I assume the file you have is a sequence of frames, not just one single frame. Then to play that, you could create a MemoryStream from your array, and on each sample request, you read 610560 bytes (12 bits per pixel for YUV420) from the stream and create a sample from that. You increase the timestamp on each sample request by a fixed amount. You do not need to set DecodeTimestamp, since you have uncompressed samples anyways. Just the one you pass to the sample constructor.

To understand your actual app scenario correctly: You have some kind of live video stream with unstable frame rate (video call or similar), right? Any you want to disaplay frames as they arrive. Then you should create a Stopwatch when the first frame arrives, and every time you deliver a new frame, you use Stopwatch.Elapsed as Timestamp. The SampleRequested event will be called much faster than your frames arrive (engine tries to buffer). You need to delay returning the sample until your next frame is available. Consider using BlockingCollection or similar producer/consumer list.

Also, be sure to set IsLive=true on the MSS and set BufferTime to zero.

**Reply by MayankKr10** [reply]
The solution for test app worked fine but I am facing trouble in integrating the solution in the actual app.

Firstly, I am receiving the frames (byte array) through an event, say OnFrameReceived. How do I delay returning the sample until next frame is received? 
Currently I am trying to save the last received frame and return the sample using this saved frame until the next frame is received. I know this is not efficient and will use up a lot of memory, and thus want to avoid it

Secondly, I assume TimeStamp in MSS is for individual frames. So, rather than starting a new Stopwatch on receiving first frame, I have to reset this Stopwatch every time a new frame is received using Stopwatch.StartNew() - is this correct?

For reference, as you had suggested, I am setting IsLive = true and BufferTime to zero for the MSS:
```
VideoStreamDescriptor descriptor = new VideoStreamDescriptor(VideoEncodingProperties.CreateUncompressed(MediaEncodingSubtypes.Iyuv, 848, 480));
                mss = new MediaStreamSource(descriptor)
                {
                    IsLive = true,
                    BufferTime = new TimeSpan(0)
                };
                mss.SampleRequested += Mss_SampleRequested;
                VideoPlayer.MediaPlayer.Source = MediaSource.CreateFromMediaStreamSource(mss);
```

**Reply by lukasf** [reply]
The timestamp is the absolute position of a frame in the video stream. So you must keep the Stopwatch running, do not restart it.

I would recommend using the AsyncProducerConsumerQueue from the Nitro.AsyncEx library. Whenever a frame arrives, create a new sample with current timestamp and add it to the asyncQueue. When the stream has stopped, call asyncQueue.CompleteAdding().

This is how the SampleRequested event handler might look like, using AsyncProducerConsumerQueue:

```
private void Mss_SampleRequested(MediaStreamSource source, MediaStreamSourceSampleRequestedEventArgs args)
{
    // First get a deferral, otherwise you cannot use await in the event handler.
    var deferral = args.Request.GetDeferral();
    try
    {
        var sample = await asyncQueue.DequeueAsync();
        args.Request.Sample = sample;
    }
    catch (InvalidOperationException)
    {
        // This occurs when CompleteAdding() was called and all samples were processed. Stream is stopped.
    }
    catch (Exception)
    {
        // Other error
    }
    // Deferral must be completed even if no sample was set. If no sample is set, playback will stop.
    deferral.Complete();
}
```

You could also use the standard BlockingCollection from System.Collections.Concurrent, but it is not nice because it will block the current thread until new samples are added. The AsyncProducerConsumerQueue supports async/await, so it won't block your threads.

**Reply by brabebhin** [reply]
I've been using https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.dataflow.bufferblock-1?view=net-7.0 for various producer/consumer applications. This is async and won't block waiting thread.

**Reply by lukasf** [reply]
Cool, thanks for that link @brabebhin! Actually, I have been looking for a good lib to build chained processing pipelines in C# since quite a while. I did not know that one. Looks really good!

**Reply by MayankKr10** [reply]
The solution you suggested has one issue we found in real-time testing - the memory utilization keeps on increasing.

Here's what we attempted on our end -
`    private async void Mss_SampleRequested(MediaStreamSource sender, MediaStreamSourceSampleRequestedEventArgs args)
        {
            var deferral = args.Request.GetDeferral();
            try
            {
                if (_stopwatch == null)
                {
                    _stopwatch = Stopwatch.StartNew();
                }

                var sample =await _consumerQ.DequeueAsync();
                args.Request.Sample = MediaStreamSample.CreateFromBuffer(sample, _stopwatch.Elapsed);
                //sample = null;
            }
            catch (Exception ex)
            {
                // Log error here
            }
            deferral.Complete();
        }`
As you can see, this is almost an exact copy of the snippet you had shared. Problem is that the memory allocated for the IBuffer in AsyncProducerConsumerQueue is not being freed even after dequeue. For now I am using GC.Collect() based on a 5 secs timer to free up the memory in heap, but I suppose it's not a good idea -
`
            _timer = new Timer
            {
                Interval = 5000
            };
            _timer.Elapsed += HandleTimerElapsed;
            _timer.Start();

        private void HandleTimerElapsed(object sender, ElapsedEventArgs e)
        {
            GC.Collect();
        }
`
Is there a better way of freeing up this memory which was allocated for the buffer earlier and reduce the overall memory?

**Reply by brabebhin** [reply]
There is no work around this. You need to decide if the GC freezes are a blocker for you. This is why we use c++ for the interop code.

**Reply by brabebhin** [reply]
The more complicated answer... The buffer doesn't get consumed immediately after you free it from the async queue. It gets consumed when the sample is rendered. But this is not the biggest problem.

The biggest problem is the way the interop works for c#. C# uses nondeterministic generational mark&sweep GC, winRT uses deterministic ref counting for memory management. In order to keep the object valid in c#, the interop layer will keep a reference alive in the winRT world, so that the winRT object doesn't self destruct. That reference isn't readily available to developers to free manually, so you have to wait for GC to collect it. The IBuffer interface doesn't provide IClosable interface, so you can't Dispose it manually, meaning the memory remains there until GC runs.

To make things even worse, the GC doesn't automatically know of the memory allocation in winRT world, so it won't know it's under memory pressure (you can add the memory pressure manually but that's error prone).

In c++ this is a non-issue, since the ref counted objects in winRT don't have hidden refs and they get to self destruct properly when they are no longer used.

**Reply by lukasf** [reply]
True. This is the reason why media engines are usually written in native code.

If the SDK allows you to provide the buffers for incoming frames, you might be able to improve things by using a buffer pool. Then instead of allocating a new byte array for every frame, you could get one from a pool (e.g. ArrayPool<byte>.Shared), and return it when the MediaStreamSample.Processed event comes. You'd need a lookup to map sample to buffer, so you know which one to return once a sample was processed.

But if this is not possible with the SDK you are using, then your only chance is to either live with GC issues, or do the sample handing in native code. Agora seems to have a native Windows SDK, so this would be possible. The rest of the app can stay in C#, only the raw data handling would have to be in a RuntimeComponent, which can then be used from the C# app, similar to how we do in our lib.

**Reply by brabebhin** [reply]
In my experience buffer pools don't help. When you marshal the IBuffer from c# to winRT, the data is actually copied by value between the two worlds. So even if you may be pooling the winRT memory, the managed memory still gets to create extra garbage, essentially a copy of the IBuffer. 

The IBuffer interface and its interop with c# is simply not designed to handle large data transfers in real time.

**Reply by lukasf** [reply]
The IBuffer interface is a pure COM interface, so it will not be copied when handing it from C# to WinRT. I am not sure about the C# array to IBuffer conversion. When looking into the the AsBuffer() extension method, it looks a bit as if this will actually wrap the existing array without copy. But only part of the implementation can be seen in the disassembly, so I might be wrong here. Still, even if the buffer is copied from C# array to IBuffer, two pools could be used, one with IBuffers and one with byte arrays, and then you copy the memory yourself and re-use both the IBuffers and the arrays. Of course, this is more work. The best solution is always to do this kind of stuff in native code and avoid all these pitfalls.
